pkg/liqo-controller-manager/namespaceOffloading-controller/namespaceoffloading_controller.go

```diff
@@ -0,0 +1,214 @@
+	client.Client
+	Scheme         *runtime.Scheme
+	LocalClusterID string
+}
+
+const (
+	namespaceOffloadingControllerFinalizer = "namespaceoffloading-controller.liqo.io/finalizer"
+)
+
+func (r *NamespaceOffloadingReconciler) getClusterIDMap(ctx context.Context, nms *mapsv1alpha1.NamespaceMapList,
+	clusterIDMap map[string]*mapsv1alpha1.NamespaceMap) error {
+	if err := r.List(ctx, nms); err != nil {
+		klog.Error(err, " --> Unable to List NamespaceMaps")
+		return err
+	}
+
+	if len(nms.Items) == 0 {
+		klog.Info("No NamespaceMaps at the moment in the cluster")
+		return nil
+	}
+
+	for i := range nms.Items {
+		clusterIDMap[nms.Items[i].Labels[liqoconst.RemoteClusterID]] = &nms.Items[i]
+	}
+	return nil
+}
+
+func (r *NamespaceOffloadingReconciler) deletionLogic(ctx context.Context,
+	noff *offv1alpha1.NamespaceOffloading, clusterIDMap map[string]*mapsv1alpha1.NamespaceMap) error {
+	klog.Infof("The NamespaceOffloading of the namespace '%s' is requested to be deleted", noff.Namespace)
+	// 1 - remove Liqo scheduling label from the associated namespace
+	if err := removeLiqoSchedulingLabel(ctx, r.Client, noff.Namespace); err != nil {
+		return err
+	}
+	// 2 - remove the involved DesiredMapping from alla NamespaceMap
+	if err := removeDesiredMappings(r.Client, noff.Namespace, clusterIDMap); err != nil {
+		return err
+	}
+	// 3 - check if all remote namespace associated with this NamespaceOffloading resource are really deleted
+	if len(noff.Status.RemoteNamespacesConditions) != 0 {
+		err := fmt.Errorf("some remote namespaces still exist")
+		klog.Info(err)
+		return err
+	}
+	// 4 - remove NamespaceOffloading controller finalizer; all remote namespaces associated with this resource
+	// have been deleted.
+	patch := noff.DeepCopy()
+	ctrlutils.RemoveFinalizer(noff, namespaceOffloadingControllerFinalizer)
+	if err := r.Patch(ctx, noff, client.MergeFrom(patch)); err != nil {
+		klog.Errorf("%s --> Unable to remove finalizer from NamespaceOffloading", err)
+		return err
+	}
+	klog.Info("Finalizer correctly removed from NamespaceOffloading")
+	return nil
+}
+
+func (r *NamespaceOffloadingReconciler) initialConfiguration(ctx context.Context,
+	noff *offv1alpha1.NamespaceOffloading) error {
+	patch := noff.DeepCopy()
+	// 1 - Add NamespaceOffloadingController Finalizer.
+	ctrlutils.AddFinalizer(noff, namespaceOffloadingControllerFinalizer)
+	// 2 - Add empty cluster selector if not specified by the user.
+	if noff.Spec.ClusterSelector.Size() == 0 {
+		noff.Spec.ClusterSelector = corev1.NodeSelector{NodeSelectorTerms: []corev1.NodeSelectorTerm{}}
+	}
+	// 3 - Add NamespaceOffloading.Status.RemoteNamespaceName.
+	if noff.Spec.NamespaceMappingStrategy == offv1alpha1.EnforceSameNameMappingStrategyType {
+		noff.Status.RemoteNamespaceName = noff.Namespace
+	} else {
+		if r.LocalClusterID == "" {
+			clusterID, err := liqoutils.GetClusterID(r.Client)
+			if err != nil {
+				return err
+			}
+			r.LocalClusterID = clusterID
+		}
```

Comment:
I would personally assume the r.LocalClusterID field to be always present, and directly use that (unless this logic is required for particular reasons)

Reply:
No this is just in case of controller failure, but I can assume the clusterID is passed again at recreation time, ty